src/utils/socket.ts

```diff
@@ -50,43 +55,116 @@ export function addListeners(
-      console.log(`received ${MALFORMED_MESSAGE_RESPONSE}`, message);
+      // this is only for validation errors, which should never occur unless
+      // the user is messing around with our internals
+      displayNotification({
+        message: 'An error has been encountered, please reload the app',
+        severity: 'error',
+      });
       history.push('/duet');
     }
   );
   socket.on(UNKNOWN_MESSAGE_RESPONSE, ({ error }: UnknownErrorResponse) => {
-    // TODO decide what to do depending on error. For now we just throw them
-    // to create
-    // TODO snackbar
-    console.log(`received ${UNKNOWN_MESSAGE_RESPONSE}`, error);
+    displayNotification({
+      message: 'An error has been encountered, please reload the app',
+      severity: 'error',
+    });
     history.push('/duet');
   });
 
   /********************* Join room ****************/
   socket.on(JOIN_ROOM_RESPONSE, (res: JoinRoomResponse) => {
     if (!res.success) {
+      if (res.code === 100) {
+        displayNotification({
+          message: "We couldn't find this room. Do you have the right code?",
+          severity: 'error',
+        });
+      } else if (res.code === 101) {
+        displayNotification({
+          message: 'This room is already full',
+          severity: 'error',
+        });
+      }
       // failed to join room, redirect back to duet page
-      // TODO snackbar
-      console.log('failed to join room', res);
       history.push('/duet');
       return;
     }
 
     const { roomInfo, playerId } = res as JoinRoomSuccessResponse;
     setRoomState(roomInfo);
+    myPlayerId = playerId;
     setPlayerId(playerId);
   });
 
   /****************** Room info updated **********************/
   socket.on(ROOM_INFO_UPDATED_NOTIFICATION, (roomInfo: RoomInfo) => {
-    setRoomState(roomInfo);
+    let message: Notification | null = null;
+    setRoomState((prevRoomInfo: RoomInfo) => {
+      if (prevRoomInfo.players.length === 0) {
+        return roomInfo;
+      }
+
+      const prevFriend = prevRoomInfo.players.find(
+        player => player.id !== myPlayerId
+      );
+
+      const curFriend = roomInfo.players.find(
+        player => player.id !== myPlayerId
+      );
+
+      if (!prevFriend && !!curFriend) {
+        message = {
+          message: 'Your partner has joined the room',
+          severity: 'info',
+        };
+      } else if (!!prevFriend && !curFriend) {
+        message = {
+          message: 'Your partner has left the room',
+          severity: 'info',
+        };
+      } else if (
+        !!prevFriend &&
+        !!curFriend &&
+        prevFriend.ready &&
+        !curFriend.ready
+      ) {
+        message = {
+          message: 'Your partner has unreadied',
+          severity: 'info',
+        };
+      } else if (
+        !!prevFriend &&
+        !!curFriend &&
+        !prevFriend.ready &&
+        curFriend.ready
+      ) {
+        message = {
+          message: 'Your partner has readied',
```

Comment:
Can we change to 'Your partner is ready now'

hmm, idk 'has readies' sounds weird

Reply:
Yeah i was trying to match the 'Your partner has unreadied', both feel really weird. I changed it to 
' Your partner is ready now'
and
'Your parter is not ready anymore'
Do they sound better?

Reply:
yup sounds better